graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,251 @@
+            schema_info, filter_infos, parameters, vertex_type_name, class_count
+        )
+
+    # TODO(bojanserafimov): If there's a many-to-one edge from A to B in the query, the
+    #                       distinct result set estimate for B cannot be greater than
+    #                       the one for A. Taking this into account would make the results
+    #                       more accurate.
+
+    return distinct_result_set_estimates
+
+
+def get_pagination_capacities(
+    schema_info: QueryPlanningSchemaInfo,
+    field_value_intervals: Dict[PropertyPath, Interval[Any]],
+    distinct_result_set_estimates: Dict[VertexPath, float],
+    query_metadata: QueryMetadataTable,
+    parameters: Dict[str, Any],
+) -> Dict[PropertyPath, float]:
+    """Get the pagination capacity for each eligible pagination field.
+
+    The pagination capacity of a field is defined as the maximum number of pages we can split
+    the query results in by adding filters on this field with some confidence that the pages
+    will have similar sizes. This reasoning is local: if a filter in a different location is
+    correlated with the values on this field, the generated pages might turn out to have
+    wildly different sizes. This problem is somewhat unavoidable.
+
+    Args:
+        schema_info: QueryPlanningSchemaInfo
+        field_value_intervals: see get_field_value_intervals
+        distinct_result_set_estimates: see get_distinct_result_set_estimates
+        query_metadata: a representation of the query
+        parameters: the query parameters
+
+    Returns:
+        The pagination capacity of each Property path
+    """
+    pagination_capacities = {}
+    for location, location_info in query_metadata.registered_locations:
+        vertex_type_name = location_info.type.name
+
+        for field_name, _ in location_info.type.fields.items():
+            key = (location.query_path, field_name)
+            if not is_meta_field(field_name):
+                if is_uuid4_type(schema_info, vertex_type_name, field_name):
+                    pagination_capacities[key] = distinct_result_set_estimates[location.query_path]
+                elif field_supports_range_reasoning(schema_info, vertex_type_name, field_name):
+                    field_value_interval = field_value_intervals.get(key, Interval(None, None))
+                    quantiles = schema_info.statistics.get_field_quantiles(
+                        vertex_type_name, field_name
+                    )
+                    if quantiles is not None:
+
+                        # Since we can't be sure the minimum observed value is the
+                        # actual minimum value, we treat values less than it as part
+                        # of the first quantile. Similarly, we treat values greater than the known
+                        # maximum as part of the last quantile. That's why we drop the minimum and
+                        # maximum observed values from the quantile list.
```

Comment:
"We treat values less than it part of the quantile". This comment doesn't make any sense. A quantile is a precise value not a collection of values. I know that this is a technicality but I think that it's really important to be as precise as possible here. I know this comment is repeated multiple times throughout the code base so feel free to modify it in those places too. I won't block you though if you don't feel it is worth it at the moment to fix this comment in the other places.

Reply:
I think that the code would be much clearer if you explained in this comment instead how you used quantiles to compute pages. Maybe through a simple example like "suppose that we have quantiles [1,2,3] then we would have 4 pages [[-inf, 1), [1,2), [2,3), [3, inf]] ". I don't know if this example is right but you probably get the point.